internal/eventstream/client.go

```diff
@@ -0,0 +1,415 @@
+	type eventCache struct {
+		SourceId int64
+		Name     string
+		Type     string
+	}
+
+	var (
+		// replayFinCh holds a reference when the Client is in the replay phase. It will be closed from the producer,
+		// startReplayWorker, when replaying is finished, which indicates the select below to switch phases. When this
+		// variable is nil, the Client is in the normal operating phase.
+		replayFinCh chan struct{}
+
+		// replayBuffer holds Event Stream events to be replayed after the replay phase has finished.
+		replayBuffer = make([]*event.Event, 0)
+		// replayCache maps eventCache derived from event.Events to API time to skip replaying outdated events.
+		replayCache = make(map[eventCache]time.Time)
+	)
+
+	// replayCacheUpdate updates the replayCache if this eventMsg seems to be the latest of its kind.
+	replayCacheUpdate := func(ev *eventMsg) {
+		evc := eventCache{ev.event.SourceId, ev.event.Name, ev.event.Type}
+		ts, ok := replayCache[evc]
+		if !ok || ev.apiTime.After(ts) {
+			replayCache[evc] = ev.apiTime
+		}
+	}
+
+	for {
+		select {
+		case <-client.Ctx.Done():
+			client.Logger.Warnw("Closing down main worker as context is finished", zap.Error(client.Ctx.Err()))
+			return
+
+		case <-client.replayPhaseRequest:
+			if replayFinCh != nil {
+				// There shouldn't be multiple concurrent startReplayWorker calls. However, technically this is possible,
+				// i.e. when the Icinga 2 API connection is flapping.
+				client.Logger.Error("Another replay phase request was sent while the Client is already replaying")
+				break
+			}
+
+			client.Logger.Debug("Dispatcher enters replay phase, starting caching Event Stream events")
+			replayFinCh = client.startReplayWorker()
+
+		case <-replayFinCh:
+			skipCounter := 0
+			for _, ev := range replayBuffer {
+				ts, ok := replayCache[eventCache{ev.SourceId, ev.Name, ev.Type}]
+				if ok && ev.Time.Before(ts) {
+					client.Logger.Debugw("Skip replaying outdated Event Stream event", zap.Stringer("event", ev),
+						zap.Time("event timestamp", ev.Time), zap.Time("cache timestamp", ts))
+					skipCounter++
+					continue
+				}
+
+				client.CallbackFn(ev)
+			}
+			client.Logger.Infow("Dispatcher leaves replay phase, returning to normal operation",
+				zap.Int("cached events", len(replayBuffer)), zap.Int("skipped events", skipCounter))
+
+			replayFinCh = nil
+			replayBuffer = make([]*event.Event, 0)
+			replayCache = make(map[eventCache]time.Time)
+
+		case ev := <-client.eventDispatcherEventStream:
+			if replayFinCh != nil {
+				replayBuffer = append(replayBuffer, ev.event)
+				replayCacheUpdate(ev)
+				break
+			}
+
+			client.CallbackFn(ev.event)
+
+		case ev := <-client.eventDispatcherReplay:
+			if replayFinCh == nil {
+				client.Logger.Errorw("Dispatcher received replay event outside of the replay phase", zap.Stringer("event", ev.event))
+				break
+			}
+
+			client.CallbackFn(ev.event)
+			replayCacheUpdate(ev)
+		}
```

Comment:
Did you consider the case where `case <-replayFinCh:` and `case ev := <-client.eventDispatcherReplay:` might become ready at the same time? I think currently they can't, but only because `client.eventDispatcherReplay` is unbuffered.

`startReplayWorker()` is already creating `replayFinCh` now, I think this would be more robust if instead of a `chan struct{}`, this was also the output stream, i.e. `chan *eventMsg` which completely replaces `client.eventDispatcherReplay`. Then with a single case for both cases that uses the `ev, ok := <- ...` receive to check whether the channel was closed.

Reply:
> Did you consider the case where `case <-replayFinCh:` and `case ev := <-client.eventDispatcherReplay:` might become ready at the same time? I think currently they can't, but only because `client.eventDispatcherReplay` is unbuffered.

Correct. Currently this should not be possible, as the channel is closed after all producing goroutines have finished.
https://github.com/Icinga/icinga-notifications/blob/500b2a974d5c03d6e811735d62d3db0f6a433d93/internal/eventstream/client.go#L283-L290

Furthermore, I'd like to keep those channels unbuffered as otherwise the producer cannot know if its message was really sent, even as this is currently not completely relevant.

> `startReplayWorker()` is already creating `replayFinCh` now, I think this would be more robust if instead of a `chan struct{}`, this was also the output stream, i.e. `chan *eventMsg` which completely replaces `client.eventDispatcherReplay`. Then with a single case for both cases that uses the `ev, ok := <- ...` receive to check whether the channel was closed.

I like this idea.

Reply:
> Furthermore, I'd like to keep those channels unbuffered as otherwise the producer cannot know if its message was really sent, even as this is currently not completely relevant.

That also wasn't intended as a suggestion to make the channels buffered. I just wanted to say that arguing why this behaves correctly requires some indirection (closing `replayFinCh` only happens after other goroutines finished, they finish only when they sent to the channel, sending to the channel requires active receiver, ...), to making this more obvious should be a good idea.